lib/types/ipblock.go

```diff
@@ -0,0 +1,201 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2020 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package types
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"net"
+	"strings"
+)
+
+// ipBlock represents a continuous segment of IP addresses
+type ipBlock struct {
+	firstIP, count *big.Int
+	ipv6           bool
+}
+
+// ipPoolBlock is almost the same as ipBlock but is put in a IPPool and knows it's start id indest
+// of it's size/count
+type ipPoolBlock struct {
+	firstIP, startIndex *big.Int
+}
+
+// IPPool represent a slice of IPBlocks
+type IPPool struct {
+	list  []ipPoolBlock
+	count *big.Int
+}
+
+func getIPBlock(s string) (*ipBlock, error) {
+	switch {
+	case strings.Contains(s, "-"):
+		return ipBlockFromRange(s)
+	case strings.Contains(s, "/"):
+		return ipBlockFromCIDR(s)
+	default:
+		if net.ParseIP(s) == nil {
+			return nil, fmt.Errorf("%s is not a valid IP, IP range or CIDR", s)
+		}
+		return ipBlockFromRange(s + "-" + s)
+	}
+}
+
+func ipBlockFromRange(s string) (*ipBlock, error) {
+	ss := strings.SplitN(s, "-", 2)
+	ip0Str, ip1Str := strings.TrimSpace(ss[0]), strings.TrimSpace(ss[1])
+	ip0, ip1 := net.ParseIP(ip0Str), net.ParseIP(ip1Str)
+	if ip0 == nil || ip1 == nil {
+		return nil, errors.New("wrong IP range format: " + s)
+	}
+	if (ip0.To4() == nil) != (ip1.To4() == nil) { // XOR
+		return nil, errors.New("mixed IP range format: " + s)
+	}
+	block := ipBlockFromTwoIPs(ip0, ip1)
+
+	if block.count.Sign() <= 0 {
+		return nil, errors.New("negative IP range: " + s)
+	}
+	return block, nil
+}
+
+func ipBlockFromTwoIPs(ip0, ip1 net.IP) *ipBlock {
+	// This code doesn't do any checks on the validity of the arguments, that should be
+	// done before and/or after it is called
+	var block ipBlock
+	block.firstIP = new(big.Int)
+	block.count = new(big.Int)
+	block.ipv6 = ip0.To4() == nil
+	if block.ipv6 {
+		block.firstIP.SetBytes(ip0.To16())
+		block.count.SetBytes(ip1.To16())
+	} else {
+		block.firstIP.SetBytes(ip0.To4())
+		block.count.SetBytes(ip1.To4())
+	}
+	block.count.Sub(block.count, block.firstIP)
+	block.count.Add(block.count, big.NewInt(1))
+
+	return &block
+}
+
+func ipBlockFromCIDR(s string) (*ipBlock, error) {
+	_, pnet, err := net.ParseCIDR(s)
```

Comment:
_, pnet, err := net.ParseCIDR(strings.TrimSpace(s))

Reply:
I don't think the trimming is needed - I just don't think people should add a random amount of additional space in their argument and then tools to just remove them for them. 

I am actually thinking if I shouldn't drop it from the places I have apparently not dropped it from :thinking: 
cc @na-- @imiric

Reply:
at least it should keep as same as ipBlockFromRange() which is using trimming.

Reply:
I don't have a strong preference either way, but we should keep it consistent, so removing trimming LGTM.